xtask/src/main.rs

```diff
@@ -0,0 +1,605 @@
+    let path = Path::new("firmware/qemu/src/bin");
+
+    let filename;
+    if release_mode {
+        filename = format!("{}.release.out", name);
+    } else {
+        filename = format!("{}.out", name);
+    }
+
+    let path = path.join(filename);
+
+    // for error context closure
+    let path_str = path.to_str().unwrap_or("(non-Unicode path)").to_string();
+
+    let content = fs::read_to_string(path)
+        .with_context(|| format!("Failed to load expected output data from {}", path_str))?;
+    Ok(content)
+}
+
+fn test_single_snapshot(name: &str, features: &str, release_mode: bool) -> Result<()> {
+    let display_name = format!(
+        "{} ({})",
+        name,
+        if release_mode { "release" } else { "dev" }
+    );
+    println!("{}", display_name);
+    let cwd_name = "firmware/qemu".to_string();
+    let mut args;
+    if release_mode {
+        args = vec!["-q", "rrb", name]
+    } else {
+        args = vec!["-q", "rb", name]
+    }
+    if !features.is_empty() {
+        args.extend_from_slice(&["--features", features]);
+    }
+
+    let actual = run_capturing_stdout(Command::new("cargo").args(&args).current_dir(cwd_name))?;
+
+    let expected = load_expected_output(name, release_mode)?;
+
+    let diff = TextDiff::from_lines(&expected, &actual);
+
+    // if anything isn't ChangeTag::Equal, print it and turn on error flag
+    let mut actual_matches_expected = true;
+    for op in diff.ops() {
+        for change in diff.iter_changes(op) {
+            let styled_change = match change.tag() {
+                ChangeTag::Delete => Some(("-", Style::new().red())),
+                ChangeTag::Insert => Some(("+", Style::new().green())),
+                ChangeTag::Equal => None,
+            };
+            if let Some((sign, style)) = styled_change {
+                actual_matches_expected = false;
+                eprint!("{}{}", style.apply_to(sign).bold(), style.apply_to(change),);
+            }
+        }
+    }
+
+    if actual_matches_expected {
+        Ok(())
+    } else {
+        Err(anyhow!("{}", display_name))
+    }
+}
+
+fn get_installed_targets() -> Result<HashSet<String>> {
+    let out = run_capturing_stdout(Command::new("rustup").args(&["target", "list"]))?;
+    let installed_marker = " (installed)";
+    let mut targets = out.lines().collect::<Vec<_>>();
+    targets.retain(|target| target.contains(installed_marker));
+    let targets: HashSet<String> = targets
+        .iter()
+        .map(|target| target.replace(installed_marker, ""))
+        .collect();
+    Ok(targets)
+}
+
+fn install_targets() -> Result<Vec<String>> {
+    let required_targets = vec![
+        "thumbv6m-none-eabi",
+        "thumbv7m-none-eabi",
+        "thumbv7em-none-eabi",
+        "thumbv8m.base-none-eabi",
+        "riscv32i-unknown-none-elf",
+    ];
+    let all_targets = required_targets
+        .iter()
+        .map(|item| item.to_string())
+        .collect::<HashSet<_>>();
```

Comment:
Is this conversion really needed? I assume this is in order to be able to use `.difference`, right?

Either we can keep `required_targets` a `let` and call `.to_string()` during initialization, or we figure out if `.difference(...)` can be called on `&str` too.

Reply:
ultimately this function needs to return a bunch of owned strings, feel free to suggest an alternative for making them. I'd aim for "less lines of code and less code in those lines", whatever satisfies that is fine with me

Reply:
See suggestion here: https://github.com/knurling-rs/defmt/pull/393#discussion_r596423372